docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,264 @@
+As noted above, this is considered a feature for diplomat, but a limitation for UniFFI.
+
+This is the problem which caused us to decide to stop working on
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) - the current world where the type universe
+is described externally doesn't have this limitation - only the UDL file needs to be parsed when
+generating the foreign bindings. The application-services team has
+concluded that none of our non-trival use-cases for UniFFI could be described using macros,
+so supporting both mechanisms is pain for no gain.
+
+As noted in #416, `wasm-bindgen` has a similarly shaped problem, and solves it by having
+the Rust macro arrange for the resulting library to have an extra data section with the
+serialized "type universe" - foreign binding generation would then read this information from the
+already built binary. This sounds more complex than the UniFFI team has appetite for at
+the current time.
+
+# Looking forward
+
+## Adapting the diplomat/[#416](https://github.com/mozilla/uniffi-rs/pull/416) model to process the entire crate?
+
+We noted that diplomat intentionally restricts where the ffi is generated,
+whereas UniFFI considers that a limitation - but what if we can teach UniFFI to process
+more of the Rust crate?
+
+It might be reasonable for the foreign bindings to know that Rust "paths" to modules which should
+be processed, and inside those modules find structs "marked up" as being used by the FFI.
+
+In other words, borrowing the example above:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+maybe can made to work, so long as we are happy to help UniFFI discover where such annotations
+may exist.
+
+A complication here is that currently UniFFI allows types defined in external crates,
+but that might still be workable - eg,
+[diplomat has an issue open to support exactly this](https://github.com/rust-diplomat/diplomat/issues/34)
+
+## Duplicating structs inside Rust
+
+When reviewing the draft of this document, @rfk noted that we are already duplicating Rust structs
+in UDL and in Rust. So instead of having:
+
+```
+// In a UDL file:
+dictionary MyFFIType {
+    i32 a;
+    bool b;
+};
+// Then in Rust:
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+we could have:
+```rust
+// In the Rust implementation, in some other module.
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+
+// And to expose it over the FFI:
+#[ffi::something]
+mod ffi {
+
+    #[ffi::magic_external_type_declaration]
+    pub struct MyFFIType {
+        pub a: i32,
+        pub b: bool,
+    }
+
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+So while we haven't exactly reduced the duplication, we have removed the UDL.
+We probably also haven't helped with documentation, because the natural location for
+the documentation of `MyFFIType` is probably at the *actual* implementation.
```

Comment:
This section is some serious food for thought.

For one, as rfk points out, we're currently duplicating the struct/function definitions in the UDL.  The way I see it: both projects have made a similar design decision.  @Manishearth describes it as "we do not want changes far away to change the FFI API".  I would maybe reword that to "the FFI should be fully defined in one place".  For UniFFI, that place is the UDL.  For diplomat, it's the ffi module.  (Multiple ffi modules complicates this picture a bit, but doesn't fundamentally change things).

However, there is one difference: in some cases you only need code inside the ffi module.  This has the potential to reduce duplication.  For example, it would be very natural to move our `Store` definitions inside the ffi module.  Also, since types defined inside the ffi module are visible to the rest of the rust code, so we could move types like `Login`, `EncrytpedLogin`, `CreditCard`, `Address`, etc. into the ffi module.  I think external data types would still need duplicate definitions, but maybe that's it.

One potential issue with this is that it couples the library with the FFI code.  But this doesn't seem to be a problem for our components.  As @markh points out, our Rust APIs exists purely to service the FFIs.

It really makes me wonder about switching from a UDL-based approach to a macro based approach.  At the start, maybe each consumer would simply refactor their current UDL file to a macro.  This shouldn't be much work, maybe we could even automate it.  After that, we have the ability to do to small refactors that eliminate the unneeded duplication.

BTW, this also could solve some documentation issues since a) we can actually see docstrings when using syn and b) if there's only 1 place where a type is defined then it's clear where to put the docstrings.

Reply:
I added a brief note in a86e9496623f9bbee77fb7b3b23e960ee47ebdc2 which doesn't capture all of this comment, but does briefly say why we should consider this more.